gene/mutate.go

```diff
@@ -0,0 +1,478 @@
+// The position of the mutation should be given in base-zero genomic coordinates.
+func PointMutation(g *Gene, genomePos int, alt dna.Base) (EffectPrediction, error) {
+	var answer EffectPrediction
+	var log diff
+	var err error
+
+	genomeIndexPos := numbers.Abs(genomePos - g.startPos) // abs needed to handle negative posStrand
+
+	// Fill log before change
+	log.genomePos = genomePos
+	log.removed = make([]dna.Base, 1)
+	log.removed[0] = g.genomeSeq[genomeIndexPos]
+	log.added = make([]dna.Base, 1)
+	log.added[0] = alt
+
+	if !g.posStrand { // so that diff can be undone by another PointMutation call
+		dna.ReverseComplement(log.removed)
+	}
+
+	if alt != dna.A && alt != dna.C && alt != dna.G && alt != dna.T {
+		return answer, errors.New("alt base must be A, C, T, or G")
+	}
+
+	if genomePos < 0 {
+		return answer, errors.New("genomePos must be positive")
+	}
+	if g.posStrand {
+		if genomePos < g.startPos {
+			return answer, errors.New("input genomePos is not in the gene")
+		}
+	} else {
+		if genomePos > g.startPos {
+			return answer, errors.New("input genomePos is not in the gene")
+		}
+		alt = dna.ComplementSingleBase(alt)
+	}
+
+	if genomeIndexPos > len(g.genomeSeq)-1 {
+		return answer, errors.New("input genomePos is not in the gene")
+	}
+
+	// if no error then make the change and append log
+	g.genomeSeq[genomeIndexPos] = alt
+	g.changeLog = append(g.changeLog, log)
+
+	cdnaIndexPos := int(g.featureArray[genomeIndexPos])
+
+	if cdnaIndexPos >= 0 { // mutation is coding
+		answer.CdnaPos = cdnaIndexPos
+		answer.AaPos = cdnaIndexPos / 3
+		codon, err := CdnaPosToCodon(g, cdnaIndexPos)
+		common.ExitIfError(err)
+		answer.AaRef = []dna.AminoAcid{dna.TranslateCodon(&codon)}
+		g.cdnaSeq[cdnaIndexPos] = alt
+		answer.AaAlt = []dna.AminoAcid{dna.TranslateCodon(&codon)}
+		if answer.AaRef[0] == answer.AaAlt[0] {
+			answer.Consequence = Silent
+		} else {
+			if answer.AaAlt[0] == dna.Stop {
+				answer.Consequence = Nonsense
+			} else if answer.AaRef[0] == dna.Stop {
+				answer.Consequence = DisruptStop
+			} else if answer.AaPos == 0 {
+				answer.Consequence = DisruptStart
+			} else {
+				answer.Consequence = Missense
+			}
+		}
+	} else { // mutation is noncoding
+		answer.CdnaPos, answer.CdnaOffset, err = GenomicPosToCdna(g, genomePos)
+		common.ExitIfError(err)
+		answer.Consequence = checkSplice(answer.CdnaOffset)
+	}
+
+	return answer, nil
+}
+
+//TODO EffectPrediction
+// Insertion adds bases to the Gene, predicts the effect, and updates the Gene struct to reflect the change.
+// The position should be the base-zero genomic coordinate of the base directly BEFORE the inserted bases.
+func Insertion(g *Gene, genomePos int, alt []dna.Base) (EffectPrediction, error) {
+	var answer EffectPrediction
+	var log diff
+	var genomeIndexPos int
+	var err error
+
+	// Fill changeLog
+	log.genomePos = genomePos
+	log.added = make([]dna.Base, len(alt))
+	copy(log.added, alt)
+
+	for _, val := range alt {
```

Comment:
This could be split out as a separate function (maybe in the dna package) that is something like:
func IsSeqOfACGT([]dna.Base) bool {